File: go/descriptor_proto_test.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

package tabnasproto

// protobuf's own descriptor.proto (test/descriptor/README.md): the one
// real-world file that names fields after nearly every keyword and
// declares `enum Edition`. TypeScript and Rust check the same facts.

import (
	"os"
	"reflect"
	"testing"
)

func TestDescriptorProto(t *testing.T) {
	src, err := os.ReadFile("../test/descriptor/descriptor.proto")
	if err != nil {
		t.Fatal(err)
	}
	fdp := mustParse(t, string(src), nil)
	if fdp.Syntax != "proto2" || fdp.Package != "google.protobuf" {
		t.Fatalf("syntax %q package %q", fdp.Syntax, fdp.Package)
	}
	var names []string
	fields, nested := 0, 0
	for _, m := range fdp.MessageType {
		names = append(names, m.Name)
		fields += len(m.Field)
		nested += len(m.NestedType)
	}
	want := []string{
		"FileDescriptorSet", "FileDescriptorProto", "DescriptorProto",
		"ExtensionRangeOptions", "FieldDescriptorProto", "OneofDescriptorProto",
		"EnumDescriptorProto", "EnumValueDescriptorProto", "ServiceDescriptorProto",
		"MethodDescriptorProto", "FileOptions", "MessageOptions", "FieldOptions",
		"OneofOptions", "EnumOptions", "EnumValueOptions", "ServiceOptions",
		"MethodOptions", "UninterpretedOption", "FeatureSet", "FeatureSetDefaults",
		"SourceCodeInfo", "GeneratedCodeInfo",
	}
	if !reflect.DeepEqual(names, want) {
		t.Fatalf("messages %v", names)
	}
	if fields != 143 || nested != 12 {
		t.Fatalf("%d fields, %d nested messages", fields, nested)
	}
	var enums []string
	for _, e := range fdp.EnumType {
		enums = append(enums, e.Name)
	}
	if !reflect.DeepEqual(enums, []string{"Edition", "SymbolVisibility"}) {
		t.Fatalf("enums %v", enums)
	}

	var fileFields []string
	for _, f := range fdp.MessageType[1].Field {
		fileFields = append(fileFields, f.Name)
	}
	if !reflect.DeepEqual(fileFields, []string{
		"name", "package", "dependency", "public_dependency", "weak_dependency",
		"option_dependency", "message_type", "enum_type", "service", "extension",
		"options", "source_code_info", "syntax", "edition",
	}) {
		t.Fatalf("FileDescriptorProto fields %v", fileFields)
	}
	for _, m := range fdp.MessageType {
		if m.Name != "FieldOptions" {
			continue
		}
		for _, f := range m.Field {
			if f.Name != "weak" {
				continue
			}
			if f.Number != 10 || f.Type != "TYPE_BOOL" || f.DefaultValue != "false" ||
				len(f.Options) != 1 || f.Options["deprecated"] != true {
				t.Fatalf("FieldOptions.weak = %+v", f)
			}
		}
	}
	edition := fdp.EnumType[0]
	if len(edition.Value) != 14 || edition.Value[0].Name != "EDITION_UNKNOWN" ||
		edition.Value[1].Number != 900 || edition.Value[2].Name != "EDITION_PROTO2" {
		t.Fatalf("Edition values %+v", edition.Value)
	}
}

File: rs/tests/descriptor_proto_test.rs
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

//! protobuf's own descriptor.proto (test/descriptor/README.md): the one
//! real-world file that names fields after nearly every keyword and
//! declares `enum Edition`. TypeScript and Go check the same facts.

use tabnas_proto::parse;

const SRC: &str = include_str!("../../test/descriptor/descriptor.proto");

#[test]
fn descriptor_proto_parses_to_the_descriptor_protoc_sees() {
    let fdp = parse(SRC, None).unwrap_or_else(|e| panic!("descriptor.proto: {e}"));
    assert_eq!(fdp.syntax.as_deref(), Some("proto2"));
    assert_eq!(fdp.package.as_deref(), Some("google.protobuf"));

    let names: Vec<&str> = fdp.message_type.iter().map(|m| m.name.as_str()).collect();
    assert_eq!(
        names,
        [
            "FileDescriptorSet",
            "FileDescriptorProto",
            "DescriptorProto",
            "ExtensionRangeOptions",
            "FieldDescriptorProto",
            "OneofDescriptorProto",
            "EnumDescriptorProto",
            "EnumValueDescriptorProto",
            "ServiceDescriptorProto",
            "MethodDescriptorProto",
            "FileOptions",
            "MessageOptions",
            "FieldOptions",
            "OneofOptions",
            "EnumOptions",
            "EnumValueOptions",
            "ServiceOptions",
            "MethodOptions",
            "UninterpretedOption",
            "FeatureSet",
            "FeatureSetDefaults",
            "SourceCodeInfo",
            "GeneratedCodeInfo",
        ]
    );
    let enums: Vec<&str> = fdp.enum_type.iter().map(|e| e.name.as_str()).collect();
    assert_eq!(enums, ["Edition", "SymbolVisibility"]);
    let fields: usize = fdp.message_type.iter().map(|m| m.field.len()).sum();
    let nested: usize = fdp.message_type.iter().map(|m| m.nested_type.len()).sum();
    assert_eq!((fields, nested), (143, 12));

    let file = &fdp.message_type[1];
    let file_fields: Vec<&str> = file.field.iter().map(|f| f.name.as_str()).collect();
    assert_eq!(
        file_fields,
        [
            "name",
            "package",
            "dependency",
            "public_dependency",
            "weak_dependency",
            "option_dependency",
            "message_type",
            "enum_type",
            "service",
            "extension",
            "options",
            "source_code_info",
            "syntax",
            "edition",
        ]
    );
    let weak = fdp
        .message_type
        .iter()
        .find(|m| m.name == "FieldOptions")
        .and_then(|m| m.field.iter().find(|f| f.name == "weak"))
        .expect("FieldOptions.weak");
    assert_eq!(weak.number, 10.0);
    assert_eq!(weak.default_value.as_deref(), Some("false"));
    assert_eq!(serde_json::to_value(weak).unwrap()["type"], "TYPE_BOOL");
    assert_eq!(
        serde_json::to_value(weak).unwrap()["options"]["deprecated"],
        true
    );

    let edition = &fdp.enum_type[0];
    assert_eq!(edition.value.len(), 14);
    assert_eq!(edition.value[0].name, "EDITION_UNKNOWN");
    assert_eq!(edition.value[1].number, 900.0);
    assert_eq!(edition.value[2].name, "EDITION_PROTO2");
}
